keps/sig-api-machinery/0034-20190129-kflag.md

```diff
@@ -0,0 +1,188 @@
+  it may need to apply values to each independently.
+* Components must be able to specify custom behavior for merging values during application.
+  Feature gates, for example, are merged piecemeal between the command-line and config files, with
+  gates specified on the command-line taking precedence.
+* Application of flag values should only consider the flags specified on the command line. It should
+  be careful, for example, to not overwrite values loaded from config files with the default values
+  of omitted flags.
+
+
+### Implementation Details
+
+Fundamentally, separating parsing from application requires that we have a scratch space to parse
+the initial values into, and a way to map that scratch space onto target objects.
+
+To maintain type safety, this proposal recommends a two-layered approach, where kflag handles the
+common types, and components handle the component-specific types:
+* kflag Implements:
+  * Statically typed flag registration helpers that allocate scratch space.
+  * Statically typed helpers to apply the values of parsed flags to arbitrary targets.
+* Components Implement:
+  * Aggregate flag registration functions, which use kflag helpers to register groups of flags the
+    component cares about.
+  * Aggregate flag application functions, which use kflag helpers to apply a subset of flag values
+    to an arbitrary structure determined by the component.
+
+This proposal recommends the approach in this example PR:
+* https://github.com/kubernetes/kubernetes/pull/73494
+
+See also @sttts's prototype for an alternative approach that predates this KEP:
+* https://github.com/kubernetes/kubernetes/pull/72037
+
+### Risks and Mitigations
+
+#### Maintenance
+The kflag package is a wrapper around an existing package, pflag.
+The pflag package has a stable API, so we would at most expect to extend kflag when pflag is
+extended, but only in the event that we really need that extension. We expect this to be infrequent.
+Given that kflag should make components easier to maintain (it reduces the complexity of the
+bootstrap process by making the flag precedence implementation easier to use and reason about),
+the maintenance of kflag itself is likely time well spent.
+
+#### Migrating existing components
+This should amount to an internal refactoring for existing components, with no external changes in
+behavior (if we do change external-facing behavior, it's a bug). Some components may have better
+test coverage than others, or unique features (like dynamic Kubelet config), so we will have to be
+careful to test properly on a component-by-component basis as we migrate components to use kflag.
+
+## Graduation Criteria
+
+* kflag implements helpers for the same set of types as pflag (or at least all the types used
+  by core Kubernetes components).
+* (optional) kflag implements the helpers that currently exist in k8s.io/apiserver/pkg/util/flag,
```

Comment:
In kubeadm this package is used in a couple of places. It will be beneficial for us if we don't import stuff from `k8s.io/apiserver/*` just because of a few helpers here and there. Hence, I think of this as not very optional (especially as we get a lot of pressure from external components, mainly cluster API, for kubeadm to drop more unnecessary deps and become slimmer).

Reply:
compare https://github.com/kubernetes/enhancements/pull/764#pullrequestreview-198009976. We did not state it here explicitly, but the new code will not go to k8s.io/apiserver. One of the motivations behind k8s.io/component-base was exactly that that k8s.io/apiserver is not the right location for api independent code.

Reply:
Yes, in the case of those flag helpers that currently live in apiserver, we'd move them into kflag rather than have kflag depend on apiserver.

Reply:
Yep, that's why I was bit concerned about the `optional` word there. Nothing to worry about then. Thanks folks!

Reply:
Yeah, the "optional" is more about whether we'd initially bother to provide native helpers in kflag, or just continue to have components use the generic `fs.Var` style registration for more advanced types. Since it's not that hard to implement the extra helpers, I lean towards doing so.